procrustes/kopt.py

```diff
@@ -35,148 +34,174 @@
+    if p2.ndim != 2 or p2.shape[0] != p2.shape[1]:
+        raise ValueError(f"Argument p2 should be a square array. Given p2 shape={p2.shape}")
+
+    # check whether p1 & p2 are valid permutation matrices
+    if not np.all(np.logical_or(p1 == 0, p1 == 1)):
+        raise ValueError("Elements of permutation matrix p1 can only be 0 or 1.")
+    if not np.all(np.logical_or(p2 == 0, p2 == 1)):
+        raise ValueError("Elements of permutation matrix p2 can only be 0 or 1.")
+
+    if np.all(np.sum(p1, axis=0) != 1) or np.all(np.sum(p1, axis=1) != 1):
+        raise ValueError("Sum over rows or columns of p1 matrix isn't equal 1.")
+    if np.all(np.sum(p2, axis=0) != 1) or np.all(np.sum(p2, axis=1) != 1):
+        raise ValueError("Sum over rows or columns of p2 matrix isn't equal 1.")
+
+    # check k
+    if k < 2 or not isinstance(k, (int, np.integer)):
+        raise ValueError(f"Argument k={k} must be a integer greater than 1. Give type {type(k)}")
+    if k > max(p1.shape[0], p2.shape[0]):
+        raise ValueError(f"Argument k={k} is not smaller than {max(p1.shape[0], p2.shape[0])}.")
+
+    # compute initial value of the objective function & assign best P1 & P2 matrices
+    best_f = fun(p1, p2)
+    best_p1, best_p2 = np.copy(p1), np.copy(p2)
+
+    # swap rows and columns until the permutation matrix is not improved
+    search = True
+    while search:
+        search = False
+        # make sure p1 & p2 guesses are the best permutation matrices found thus far
+        p1, p2 = np.copy(best_p1), np.copy(best_p2)
+        for perm1 in it.permutations(np.arange(p1.shape[0]), r=int(min(k, p1.shape[0]))):
+            comb1 = tuple(sorted(perm1))
+            for perm2 in it.permutations(np.arange(p2.shape[0]), r=int(min(k, p2.shape[0]))):
+                comb2 = tuple(sorted(perm2))
+                if not (perm1 == comb1 and perm2 == comb2):
+                    # permute rows of matrix P1
+                    perm_p1 = np.copy(p1)
+                    perm_p1[comb1, :] = perm_p1[perm1, :]
+                    # permute rows of matrix P2
+                    perm_p2 = np.copy(p2)
+                    perm_p2[comb2, :] = perm_p2[perm2, :]
+                    # compute objective function for permuted P matrix & compare
+                    perm_f = fun(perm_p1, perm_p2)
+                    if perm_f < best_f:
+                        best_p1, best_p2, best_f = perm_p1, perm_p2, perm_f
+                        # set search=True to keep permuting the new best_p1 & best_p2 unless this
+                        # is already an exhaustive search
+                        search = bool(k < max(p1.shape[0], p2.shape[0]))
+                        # check whether perfect permutation matrix is found
+                        # TODO: smarter threshold based on norm of matrix
+                        if best_f <= 1.0e-8:
+                            return best_p1, best_p2, best_f
```

Comment:
Since the number of permutation matrices is finite, it wasn't obvious to me that this function has to be non-negative. I would either add that it is non-negative to the document or remove it.